src/main/java/de/zalando/aruha/nakadi/service/EventStreamCoordinator.java

```diff
@@ -0,0 +1,174 @@
+        // add new client to topology and invoke rebalance
+        final String newClientId = subscriptionRepository.generateNewClientId();
+        subscriptionRepository.addClient(subscriptionId, newClientId);
+        rebalanceSubscriptionIfNeeded(subscriptionId);
+
+        // get partitions and cursors for this new client
+        final Topology topology = subscriptionRepository.getTopology(subscriptionId);
+        final int newClientIndex = topology.getClientIndex(newClientId).get();
+        final int clientsNum = topology.getClientIds().size();
+
+        final List<TopicPartition> partitionsForClient = partitionDistributor.getPartitionsForClients(subscriptionId,
+                ImmutableList.of(newClientIndex), clientsNum).get(newClientIndex);
+
+        final List<Cursor> cursors = partitionsForClient
+                .stream()
+                .map(tp -> subscriptionRepository.getCursor(subscriptionId, tp.getTopic(), tp.getPartition()))
+                .collect(Collectors.toList());
+
+        // create the event stream itself
+        final EventConsumer eventConsumer = topicRepository.createEventConsumer(cursors);
+        final EventStream eventStream = new EventStream(eventConsumer, outputStream, config, cursors);
+        eventStream.setClientId(newClientId);
+        eventStream.setSubscriptionId(subscriptionId);
+        eventStreams.add(eventStream);
+        return eventStream;
+    }
+
+    /**
+     * Removes finished stream and invokes rebalance of this stream subscription
+     *
+     * @param eventStream the stream to remove
+     */
+    public void removeEventStream(final EventStream eventStream) {
+        eventStreams.remove(eventStream);
+        subscriptionRepository.removeClient(eventStream.getSubscriptionId(), eventStream.getClientId());
+        rebalanceSubscriptionIfNeeded(eventStream.getSubscriptionId());
+    }
+
+    /**
+     * Checks if rebalance is needed for subscriptions running on this Nakadi instance
+     * todo: in general this scheduled task should be probably replaced with watching appropriate zookeeper nodes
+     */
+    @Scheduled(fixedRate = SUBSCRIPTION_REBALANCE_CHECK_PERIOD_IN_MS)
+    private void rebalanceSubscriptions() {
+        try {
+            eventStreams
+                    .stream()
+                    .map(EventStream::getSubscriptionId)
+                    .distinct()
+                    .forEach(this::rebalanceSubscriptionIfNeeded);
+        }
+        catch (Exception e) {
+            LOG.error("Error occurred during subscriptions rebalance", e);
+        }
+    }
+
+    /**
+     * If new topology for this subscription found - rebalances according to new topology
+     *
+     * @param subscriptionId id of subscription to check for rebalance
+     */
+    private void rebalanceSubscriptionIfNeeded(final String subscriptionId) {
+        final Topology topology = subscriptionRepository.getTopology(subscriptionId);
```

Comment:
I checked again possible scenarios in my head and I think if even some clients start for subscription at the same time - it will not be a problem as a scheduled task will rebalance anyway to the latest version of topology.